[hj235] iP

hj235 opened the pull request:
# Bob
>"In the one and only true way. The object-oriented version of 'Spaghetti code' is, of course, 'Lasagna code'. (Too many layers)." - Roberto Waltman ([source](https://medium.com/statuscode/the-12-most-retweeted-programming-quotes-2b039c45ca39#:~:text=A%20quote%20from%20Roberto%20Waltman,Too%20many%20layers.%E2%80%9D))

Bob frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ ***SUPER* FAST** to use

All you need to do is,

1. download it from [here](https://github.com/hj235/ip/releases/tag/v0.1-beta).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you :eggplant:

And it is **FREE!**

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

<hr>

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

hj235 committed:
Level-0: Rename, Greet, Exit

hj235 committed:
Level-1: Echo

hj235 committed:
Create lineFormat method

hj235 committed:
Level-2: Add, List

hj235 committed:
Created Task and TaskList classes

hj235 committed:
Level-3: Mark as Done

hj235 committed:
moved control flow into switch statement

hj235 committed:
refactored code in preparation for later levels

hj235 committed:
Level-4: ToDo, Event, Deadline

hj235 committed:
removed a comment

hj235 committed:
A-TextUiTesting: Automated Text UI Testing

hj235 committed:
Level-5: Handle Errors

hj235 committed:
Bug fixes

hj235 committed:
Bug fix for mark and unmark instructions

hj235 committed:
Level-6: Delete

hj235 committed:
fixed bugs in tests and changed Bob.splitInput

hj235 committed:
Level-7: Save

hj235 committed:
Merge branch 'branch-Level-7'

hj235 committed:
resolve merge conflicts

hj235 committed:
Dates and OOP with comments

hj235 committed:
Add dates, times and more OOP

hj235 committed:
Move all classes into bob package

hj235 committed:
Switch date names

hj235 committed:
Merge branch 'branch-Level-8'

hj235 committed:
Merge branch 'add-gradle-support'

hj235 committed:
handle creation of new save file

hj235 committed:
Add JUnit Tests

hj235 committed:
Set up build.gradle file for building jar file

hj235 committed:
Add JavaDocs

hj235 committed:
Merge branch 'branch-A-JavaDoc'

hj235 committed:
Follow Coding standard

hj235 committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/bob/DeadlineCommand.java

hj235 committed:
Add Find Command

hj235 committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/bob/FindCommand.java
#	src/main/java/bob/TaskList.java

Al-ez commented on src/main/java/bob/Bob.java:
```diff
@@ -0,0 +1,58 @@
+     */
+    public void run() {
+        ui.showWelcome();
+        boolean isExit = false;
```
nit: isExit sounds a little grammatically wrong, maybe use isExited?

Al-ez commented on src/main/java/bob/TaskList.java:
```diff
@@ -0,0 +1,130 @@
+        this.tasklist = new ArrayList<>();
+    }
+
+    public ToDo todo(String name) {
```
nit: i suggest using verbs as method names (from textbook), maybe AddToDo or something like CreateToDo

Al-ez commented on src/main/java/bob/TaskList.java:
```diff
@@ -0,0 +1,130 @@
+        return todo;
+    }
+
+    public Deadline deadline(String name, String by) {
```
nit: i suggest using verbs as method names (from textbook), maybe AddDeadline or something like CreateDeadline

Al-ez commented on src/main/java/bob/TaskList.java:
```diff
@@ -0,0 +1,130 @@
+        return deadline;
+    }
+
+    public Event event(String name, String from, String to) {
```
nit: i suggest using verbs as method names (from textbook), maybe AddEvent or something like CreateEvent

Al-ez reviewed (commented):
Looks pretty good, maybe just a few nit comments :) 😹

hj235 committed:
Set up build.gradle for JavaFX

hj235 committed:
Configure checkstyle for gradle

hj235 committed:
Merge branch 'branch-A-CheckStyle'

hj235 committed:
Correct code style errors in main

hj235 committed:
change commands to return output when executed

hj235 committed:
create GUI for Bob

hj235 committed:
work on exiting GUI through bye command

hj235 committed:
Merge branch 'branch-Level-10'

jereeemyyyy commented on src/main/java/bob/DeadlineCommand.java:
```diff
@@ -0,0 +1,57 @@
+package bob;
```
I noticed that all the code files are packaged under "bob".  Perhaps create separate directories/packages to group related code files together?

jereeemyyyy commented on src/main/java/bob/DeleteCommand.java:
```diff
@@ -0,0 +1,57 @@
+        "delete"
+    };
+    protected static int paramCount = 1;
+    protected static String identifier = "delete";
```
Since `identifier` is a constant, it should be in uppercase, `IDENTIFIER` according to the coding standard for this course!

jereeemyyyy commented on src/main/java/bob/Parser.java:
```diff
@@ -0,0 +1,145 @@
+        int[] indexes = new int[splits.length + 1];
+        indexes[splits.length] = input.length;
+
+        // find index of each split/parameter in the input array
+        for (int i = 0; i < splits.length; i++) {
+            int index = Parser.findIndex(input, splits[i], 0);
+            indexes[i] = index;
+            result[i] = "";
+        }
+
+        // detect if any parameters are missing
+        for (int i = 0; i < splits.length; i++) {
+            if (indexes[i] < 0 || indexes[i + 1] - indexes[i] <= 1) {
+                throw new MissingParamException(splits[i]);
+            }
+        }
+
+        for (int i = 0; i < splits.length; i++) {
+            StringBuilder arg = new StringBuilder(input[indexes[i] + 1]);
+            for (int j = indexes[i] + 2; j < indexes[i + 1]; j++) {
+                arg.append(' ');
+                arg.append(input[j]);
+            }
+            result[i] = arg.toString();
+        }
+
+        if (result.length < splitCount) {
+            throw new MissingParamException("mark");
+        } else if (result.length > splitCount) {
+            throw new ExtraParamException("mark");
+        }
+
+        return result;
+    }
```
I like how there is a separation between the different logics! Looks clean 😺

jereeemyyyy commented on src/main/java/bob/Task.java:
```diff
@@ -0,0 +1,45 @@
+        this.completed = true;
+    }
+
+    public void uncomplete() {
```
Should this be written as `unComplete()` instead?

jereeemyyyy commented on src/main/java/bob/TaskList.java:
```diff
@@ -0,0 +1,130 @@
+        return event;
+    }
+
+    public void mark(int idx) {
+        if (idx < 1 || idx > this.tasklist.size()) {
+            throw new TaskIndexException(String.valueOf(idx));
+        }
+
+        this.tasklist.get(idx - 1).complete();
+    }
+
+    public void unmark(int idx) {
+        if (idx < 1 || idx > this.tasklist.size()) {
+            throw new TaskIndexException(String.valueOf(idx));
+        }
+
+        this.tasklist.get(idx - 1).uncomplete();
+    }
+
+    public Task delete(int idx) {
+        if (idx < 1 || idx > this.tasklist.size()) {
+            throw new TaskIndexException(String.valueOf(idx));
+        }
+
+        Task deleted = this.tasklist.get(idx - 1);
+        this.tasklist.remove(idx - 1);
+        return deleted;
+    }
```
It would be clearer if you could specify exactly what the methods are doing. For example, `mark()` could be `markTask()` etc.

jereeemyyyy reviewed (commented):
Generally, it was a well-written piece of code that was easy to read and understand! LGTM, just some nits here and there to fix! 😆

hj235 committed:
Remove 'identifier' fields in Command subclasses

Command subclasses: Remove 'identifier' instance fields.

These fields are not being used in the current version.

Let's remove these fields from the Command subclasses.

The unused fields can make the code less readable.

hj235 committed:
Add java Assertions

Parser class methods are making some assumptions about inputs.

These assumptions could lead to bugs in the future, and bugs in the
runtime environment.

Let's define assumptions about the program state so the runtime
environment can verify them.

Using java assertions is preferable over JUnit assertions here as java
assertions are present within the functional code instead of the test
code.

hj235 committed:
Refactor: extract variable & decompose conditional

Some classes consisted of complicated inline variables and
conditionals.

These could make the code less readable and increase the workload
required for future development to take place.

Let's extract explaining variables and decompose complicated
conditionals into separate variables with explanatory names.

This increases the readability of the code as the names can shed some
light onto the purpose of these lines of code.

hj235 committed:
Merge pull request #1 from hj235/branch-A-Assertions

Add java assertions

hj235 committed:
Merge pull request #2 from hj235/branch-A-CodeQuality

Refactor: extract variable & decompose conditional

hj235 committed:
add undo command

hj235 committed:
Merge pull request #3 from hj235/branch-BCD-Extension

add undo command

hj235 committed:
Create file path if non-existent

hj235 committed:
Reorganise classes into packages

hj235 committed:
Close JavaFX App on 'bye' input

hj235 committed:
Refactor Parser class: extract functions

hj235 committed:
Improve Code Styling

hj235 committed:
Add css files

hj235 committed:
Merge branch 'branch-A-BetterGui'

hj235 committed:
Add exception for invalid date

hj235 committed:
Add product screenshot

hj235 committed:
Update user guide
